x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/RateLimiter.java

```diff
@@ -0,0 +1,148 @@
+        this.sleeper = Objects.requireNonNull(sleeper);
+        this.clock = Objects.requireNonNull(clock);
+        nextTokenAvailability = Instant.MIN;
+        setRate(accumulatedTokensLimit, tokensPerTimeUnit, unit);
+    }
+
+    public final synchronized void setRate(double newAccumulatedTokensLimit, double newTokensPerTimeUnit, TimeUnit newUnit) {
+        Objects.requireNonNull(newUnit);
+
+        if (newAccumulatedTokensLimit < 0) {
+            throw new IllegalArgumentException("Accumulated tokens limit must be greater than or equal to 0");
+        }
+
+        if (Double.isInfinite(newAccumulatedTokensLimit)) {
+            throw new IllegalArgumentException(
+                Strings.format("Accumulated tokens limit must be less than or equal to %s", Double.MAX_VALUE)
+            );
+        }
+
+        if (newTokensPerTimeUnit <= 0) {
+            throw new IllegalArgumentException("Tokens per time unit must be greater than 0");
+        }
+
+        if (newTokensPerTimeUnit == Double.POSITIVE_INFINITY) {
+            throw new IllegalArgumentException(Strings.format("Tokens per time unit must be less than or equal to %s", Double.MAX_VALUE));
+        }
+
+        accumulatedTokens = Math.min(accumulatedTokens, newAccumulatedTokensLimit);
+
+        accumulatedTokensLimit = newAccumulatedTokensLimit;
+
+        var unitsInNanos = newUnit.toMicros(1);
+        tokensPerMicros = newTokensPerTimeUnit / unitsInNanos;
+        assert Double.isInfinite(tokensPerMicros) == false : "Tokens per microsecond should not be infinity";
+
+        accumulateTokens();
+    }
+
+    /**
+     * Causes the thread to wait until the tokens are available
+     * @param tokens the number of items of work that should be throttled, typically you'd pass a value of 1 here
+     * @throws InterruptedException _
+     */
+    public void acquire(int tokens) throws InterruptedException {
+        if (tokens <= 0) {
+            throw new IllegalArgumentException("Requested tokens must be positive");
+        }
+
+        double microsToWait;
+        synchronized (this) {
+            accumulateTokens();
+            var accumulatedTokensToUse = Math.min(tokens, accumulatedTokens);
+            var additionalTokensRequired = tokens - accumulatedTokensToUse;
+            microsToWait = additionalTokensRequired / tokensPerMicros;
+            accumulatedTokens -= accumulatedTokensToUse;
+            nextTokenAvailability = nextTokenAvailability.plus((long) microsToWait, ChronoUnit.MICROS);
+        }
+
+        sleeper.sleep((long) microsToWait);
+    }
+
+    private void accumulateTokens() {
+        var now = Instant.now(clock);
+        if (now.isAfter(nextTokenAvailability)) {
+            var elapsedTimeNanos = microsBetweenExact(nextTokenAvailability, now);
+            var newTokens = tokensPerMicros * elapsedTimeNanos;
+            accumulatedTokens = Math.min(accumulatedTokensLimit, newTokens);
```

Comment:
Should this include the previously accumulated tokens? 

```suggestion
            accumulatedTokens = Math.min(accumulatedTokensLimit, accumulatedTokens + newTokens);
```

Reply:
Yep thanks for that.